File: src/main/java/no/nav/aap/api/sts/STSConfig.java
package no.nav.aap.api.sts;

import no.nav.aap.api.config.AbstractRestConfig;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.boot.context.properties.ConstructorBinding;
import org.springframework.boot.context.properties.bind.DefaultValue;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.web.reactive.function.BodyInserters.FormInserter;
import org.springframework.web.util.UriBuilder;

import java.net.URI;
import java.time.Duration;

import static com.nimbusds.oauth2.sdk.GrantType.CLIENT_CREDENTIALS;
import static org.springframework.web.reactive.function.BodyInserters.fromFormData;

@ConfigurationProperties(prefix = "sts")
public class STSConfig extends AbstractRestConfig {

    private static final String DEFAULT_BASE_URI = "http://must.be.set";
    private static final String GRANT_TYPE = "grant_type";
    private static final String DEFAULT_PATH = "/rest/v1/sts/token";
    private static final String DEFAULT_SLACK = "20s";
    private static final String PING_PATH = ".well-known/openid-configuration";
    private static final String SCOPE = "scope";
    private final String username;
    private final String password;
    private final Duration slack;
    private final String stsPath;

    @ConstructorBinding
    public STSConfig(@DefaultValue(DEFAULT_BASE_URI) URI baseUri,
                     @DefaultValue(DEFAULT_SLACK) Duration slack, String username, String password,
                     @DefaultValue(PING_PATH) String pingPath, @DefaultValue("true") boolean enabled,
                     @DefaultValue(DEFAULT_PATH) String stsPath) {
        super(baseUri, pingPath, enabled);
        this.stsPath = stsPath;
        this.slack = slack;
        this.username = username;
        this.password = password;
    }

    public String getUsername() {
        return username;
    }

    public String getPassword() {
        return password;
    }

    public Duration getSlack() {
        return slack;
    }

    public String getStsPath() {
        return stsPath;
    }

    URI getStsURI(UriBuilder b) {
        return b.path(stsPath)
                .build();
    }

    FormInserter<String> stsBody() {
        var m = new LinkedMultiValueMap<String, String>();
        m.add(GRANT_TYPE, CLIENT_CREDENTIALS.getValue());
        m.add(SCOPE, "openid");
        return fromFormData(m);
    }

    @Override
    public String toString() {
        return getClass().getSimpleName() + "[username=" + username + ", password=" + password + ", slack=" + slack
                + ", stsPath=" + stsPath + "]";
    }

}

File: src/main/kotlin/no/nav/aap/api/oppslag/system/STSConfig.kt
package no.nav.aap.api.oppslag.system

import com.nimbusds.oauth2.sdk.GrantType.CLIENT_CREDENTIALS
import no.nav.aap.rest.AbstractRestConfig
import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.boot.context.properties.ConstructorBinding
import org.springframework.boot.context.properties.bind.DefaultValue
import org.springframework.util.LinkedMultiValueMap
import org.springframework.web.reactive.function.BodyInserters.FormInserter
import org.springframework.web.reactive.function.BodyInserters.fromFormData
import org.springframework.web.util.UriBuilder
import java.net.URI
import java.time.Duration


const val  GRANT_TYPE = "grant_type"
const val DEFAULT_PATH = "/rest/v1/sts/token"
const val DEFAULT_SLACK = "20s"
const val PING_PATH = ".well-known/openid-configuration"
const val SCOPE = "scope"
//@ConfigurationProperties(prefix = "sts")
//@ConstructorBinding
data class STSConfig( val base: URI,
                @DefaultValue(DEFAULT_SLACK) val slack: Duration,
                val username: String,
                val password: String,
                @DefaultValue(PING_PATH) val ping: String,
                @DefaultValue("true") val enabled: Boolean,
                @DefaultValue(DEFAULT_PATH) private val stsPath:  String) : AbstractRestConfig(base,ping,enabled) {

    fun stsBody(): FormInserter<String> {
        val m = LinkedMultiValueMap<String, String>()
        m.add(GRANT_TYPE, CLIENT_CREDENTIALS.value)
        m.add(SCOPE, "openid")
        return fromFormData(m)
    }
    fun stsURI(b: UriBuilder) = b.path(stsPath).build()
}